modelling/src/main/java/org/axonframework/modelling/StateManager.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2010-2025. Axon Framework
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.modelling;
+
+import jakarta.annotation.Nonnull;
+import org.axonframework.messaging.unitofwork.ProcessingContext;
+import org.axonframework.modelling.repository.AsyncRepository;
+import org.axonframework.modelling.repository.ManagedEntity;
+
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The {@code StateManager} enables applications to load entities based on the type of the entity and an id, and to
+ * persist them. Only one {@link AsyncRepository} or load and save function can be registered for a type, as this is
+ * considered their identity.
+ * <p>
+ * Entities are registered by their type, and can be registered by either:
+ * <ul>
+ *     <li>Registering an {@link AsyncRepository} for a certain type, or</li>
+ *     <li>Registering a load and save function for a certain type.</li>
+ * </ul>
+ * <p>
+ * In the latter case, the state manager will create a {@link SimpleEntityAsyncRepository} for the given type with
+ * the given load and save functions.
+ *
+ * @author Mitchell Herrijgers
+ * @since 5.0.0
+ */
+public interface StateManager {
+
+    /**
+     * Registers an {@link AsyncRepository} for state type {@code T} with id of type {@code I}. Only one
+     * {@link AsyncRepository} can be registered for a type.
+     *
+     * @param idType     The type of the identifier.
+     * @param stateType  The type of the state.
+     * @param repository The {@link AsyncRepository} to use for loading state.
+     * @param <I>        The type of id.
+     * @param <T>        The type of state.
+     * @return This instance for fluent interfacing.
+     */
+    <I, T> StateManager register(
```

Comment:
The `StateManager` is the object we insert into the `StatefulCommandHandler`. Hence, users are able to register new repositories in message handlers. I am wondering whether we should be careful about this or not.

Reply:
Split this in to a `StateManager.Builder`, that checks for conflicts of entities and ids,  that on `.build()` makes a `StateManager` that is immutable during runtime of the application.
In short: Moved registration to configuration time